Sources/NIO/SelectableEventLoop.swift

```diff
@@ -146,13 +146,28 @@ internal final class SelectableEventLoop: EventLoop {
         }
     }
 
+    public var safeToExecute: Bool {
+        if self.inEventLoop {
+            return true
+        }
```

Comment:
I don't think this escape hatch is right. We can be in the event loop but at a stage where further enqueues of work will fail. In fact, the obvious situation where that happens is when the event loop is shut down. At that point we'll dequeue all pending tasks that have not yet executed and fail them. Failing those tasks dispatches callbacks, and those callbacks may try to call back around to add work to the task queue. That's a bad pattern, and we should catch it, but we won't if this hook is here.

I think we only want the external state check.

Reply:
Actually, I think we can do better.

The `MultithreadedEventLoopGroup` currently uses a `_scheduledTasks` `PriorityQueue` to store pending work items. This queue is guarded by `_tasksLock`. I propose that we wrap the `PriorityQueue` in a data structure that allows us to "scribble" it: to mark it permanently as unusable once we're sure we aren't going to touch it again.

We can do this just by making the `PriorityQueue` optional, I think, and force unwrapping it whenever we try to modify it. That will allow us to catch the case where tasks are enqueued too early. It doesn't give a good error message, but that might be fine.

If we really want a good error message, a custom enum that is basically an optional with the method we want would also work.

This avoids the need for a new computed property, and also avoids synchronisation concerns: we don't have to worry about communication between two locks. It also allows us to use the point in the actual code where we know no further tasks can execute (the `defer` block at the top of `func run()`) to guarantee that outcome.

Reply:
Ah, makes sense!  Looking into it.

Reply:
Ok, pushed up a new approach with _scheduledTasks as an optional

Reply:
@weissi @Lukasa I think you could break this task into 3 parts:

1) identify the minimum set of places where a check needs to be made before executing
2) identify what to query as part of the check
3) based on step 2, possibly write some data on close that can correctly fulfill the query

Since tasks can be enqueued from inside and outside the thread, I don't _think_ there is any solution to step 1 that would allow us to use just `internalState` or just `externalState` safely.

My original approach branched on `inEventLoop` (though as @Lukasa pointed out I missed a test case), alternatively, current approach adds a bit of info that can be queried from `tasksLock` and thus all cases from step 1.

Reply:
@realdoug hmm, I'm somewhat lost here. I'm pretty sure we can find out what state we're in wherever we are because:

- if we're on the `EventLoop`, we can learn that we won't be processing new tasks if the `internalState` is `noLongerRunning`
- if we're off the `EventLoop`, then we should be in `.closing`/`.closed` when we're shutting down

If we need another internal (or external) state, for example `.aboutToExitThread ` or something, then I'm happy for you to add such a state. So for example at the very point where you currently set `self._scheduledTasks = nil` you could

    assert(self.state == .noLongerRunning, "illegal state: \(self.state)")
    self.state = . aboutToExitThread

Then in `schedule0` you could

```swift
if self.inEventLoop {
    switch self.internalState {
        ...
    }
} else {
    switch self.externalState {
        ...
    }
}
```

and in those `switch`es you can then do the appropriate checks depending on whether we're inside or outside the eventloop.

Does that make sense?

Reply:
Sorry, I think I got excited and wrote too much :)  My only point was that there are two solutions: an if statement or an optional.  Using just one of the two states won't work.

I pushed up a new version w/ an if statement inside `_schedule0` based on your description.  Let me know what you think.